[docs] Melange mode in library (#6239)
ocaml/dune: doc/melange.rst

.. _melange_main:

***********************************
JavaScript Compilation With Melange
***********************************

`Melange <https://github.com/melange-re/melange>`_ is a compiler from OCaml to

> JavaScript. Unlike Js_of_ocaml, Melange works by translating OCaml compiler

Reviewer: ```suggestion
JavaScript. Unlike js_of_ocaml, Melange works by translating OCaml compiler
```
Unclear for the capitalization of this. My notes say sentence case (at beginning of sentences and in titles, but not in middle of sentences), but should the J always be capitalized?

Author: I think you are right, I checked the official documentation and it seems to follow the casing rules you mention, e.g.: http://ocsigen.org/js_of_ocaml/latest/manual/separate-compilation.

Reviewer: Great! Thanks for confirming that!

internal lambda representation to JS files. This allows to produce a single
JavaScript file from each OCaml module. Melange can be installed with

> `Opam, Esy or Nix <https://github.com/melange-re/melange#installation>`_ package

Reviewer: ```suggestion
`opam, Esy, or Nix <https://github.com/melange-re/melange#installation>`_ package
```

managers.

Compiling to JS
===============

Dune has experimental support for building Melange libraries. To use it, you
must declare the ``melange`` extension in your ``dune-project`` file:

.. code:: scheme

  (lang dune 3.5)
  (using melange 0.1)

Then, given this example:

.. code:: bash

   echo 'Js.log "hello from melange"' > foo.ml

With the following ``dune`` file:

.. code:: scheme

  (library (name foo) (modes melange))

One can then request the ``.js`` target:

.. code:: bash

   $ dune build .foo.objs/melange/foo.js
   $ node _build/default/.foo.objs/melange/foo.js
   hello from melange

At the moment, executable targets are not supported, but using explicit
targets inside libraries as shown above should enable similar results.

> Faster builds with ``subdir`` and ``dirs`` stanzas

Reviewer: ```suggestion
Faster Builds with ``subdir`` and ``dirs`` Stanzas
```

==================================================

> Melange libraries are commonly installed from the npm package repository,

Reviewer: ```suggestion
Melange libraries are commonly installed from the ``npm`` package repository,
```
Suggesting monospace because it's a repo and so it doesn't look like a typo for something else

together with other JavaScript packages. To avoid having Dune inspect
unnecessary folders in ``node_modules``, it is recommended to explicitly

> whitelist those folders that are relevant for Melange builds.

Reviewer 2: ```suggestion
include those folders that are relevant for Melange builds.
```

This can be accomplished by combining :ref:`subdir` and :ref:`dune-subdirs`

> stanzas in a ``dune`` file, and co-locate this file together with the

Reviewer: ```suggestion
stanzas in a ``dune`` file and by co-locating this file together with the
```

``node_modules`` folder. The :ref:`dune-vendored_dirs` stanza can be used as

> well to avoid warnings in Melange libraries to appear in the application build.

Reviewer: ```suggestion
well to avoid warnings in Melange libraries during the application build.
```

.. code:: scheme

  (subdir
   node_modules
   (vendored_dirs reason-react)

>    (dirs reason-react))

Reviewer 2: is `(dirs)` actually needed if `(vendored_dirs)` is passed?

Reviewer 2: ah got it, `vendored_dirs` is implicitly empty while `dirs` is implicitly everything. this is a confusing model...
